pragma-node/src/infra/repositories/entry_repository.rs

```diff
@@ -708,3 +705,225 @@ pub async fn get_ohlc(
+    raw_entries: Vec<RawMedianEntryWithComponents>,
+    pairs_ids: &[String],
+) -> Option<Vec<MedianEntryWithComponents>> {
+    if raw_entries.is_empty() {
+        return None;
+    }
+    let pairs_set: HashSet<_> = pairs_ids.iter().collect();
+    let mut found_pairs = HashSet::new();
+
+    let mut median_entries = Vec::with_capacity(raw_entries.len());
+    for raw_entry in raw_entries {
+        found_pairs.insert(raw_entry.pair_id.clone());
+
+        let median_entry = MedianEntryWithComponents::try_from(raw_entry);
+        let median_entry = match median_entry {
+            Ok(median_entry) => median_entry,
+            Err(_) => panic!("Converting raw entry to median entry failed - should not happen!"),
+        };
+
+        let num_unique_publishers = median_entry
+            .components
+            .iter()
+            .map(|c| &c.publisher)
+            .collect::<HashSet<_>>()
+            .len();
+        if num_unique_publishers < MINIMUM_NUMBER_OF_PUBLISHERS {
+            return None;
+        }
+
+        median_entries.push(median_entry);
+    }
+    if found_pairs.len() == pairs_set.len() {
+        Some(median_entries)
+    } else {
+        None
+    }
+}
+
+/// Builds a SQL query that will fetch the recent prices between now and
+/// the given interval for each unique tuple (pair_id, publisher, source)
+/// and then calculate the median price for each pair_id.
+/// We also return in a JSON string the components that were used to calculate
+/// the median price.
+fn build_sql_query_for_median_with_components(pair_ids: &[String], interval_in_ms: u64) -> String {
+    format!(
+        r#"
+            WITH last_prices AS (
+                SELECT
+                    e.pair_id,
+                    e.publisher,
+                    p.account_address AS publisher_account_address,
+                    e.source,
+                    e.price,
+                    e.timestamp,
+                    e.publisher_signature,
+                    ROW_NUMBER() OVER (PARTITION BY e.pair_id, e.publisher, e.source ORDER BY e.timestamp DESC) AS rn
+                FROM 
+                    entries e
+                JOIN
+                    publishers p ON e.publisher = p.name
+                WHERE 
+                    e.pair_id IN ({pairs_list})
+                    AND e.timestamp >= NOW() - INTERVAL '{interval_in_ms} milliseconds'
+            ),
+            filtered_last_prices AS (
+                SELECT 
+                    pair_id,
+                    publisher,
+                    publisher_account_address,
+                    source,
+                    price,
+                    timestamp,
+                    publisher_signature
+                FROM 
+                    last_prices
+                WHERE 
+                    rn = 1
+            )
+            SELECT
+                pair_id,
+                percentile_cont(0.5) WITHIN GROUP (ORDER BY price) AS median_price,
```

Comment:
maybe we should use https://docs.timescale.com/api/latest/hyperfunctions/percentile-approximation/uddsketch/

Reply:
Tried but this wasn't straightforward, errors & had a wrong median when it worked (probably bad implementation yeah). I'd say let's monitor & optimize if we have some bottleneck there